### PR TITLE
Soju: various refactors

### PR DESCRIPTION
**Pull request by laalsaas**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/soju/default.nix [hunk 1]
```diff
@@ -18,13 +18,6 @@ buildGoModule rec {
 
   vendorHash = "sha256-iT/QMm6RM6kvw69Az+aLTtBuaCX7ELAiYlj5wXAtBd4=";
 
-  subPackages = [
-    "cmd/soju"
-    "cmd/sojuctl"
-    "contrib/migrate-db"
-    "contrib/znc-import"
-  ];
-
```
Comment (malte-v) [inline pkgs/applications/networking/soju/default.nix:27]: is this superfluous?

**Comment by laalsaas** [comment]
Not only that, it prevebts `cmd/sojudb` from getting built, wich is necessary to initialize the db.

**Comment by malte-v** [comment]
ah ok👍 LGTM then

**Review by malte-v — APPROVED** [review]

File: pkgs/applications/networking/soju/default.nix (hunk 1 above)
Comment (azahi) [reply pkgs/applications/networking/soju/default.nix:27]: Please add `cmd/sojudb` and `contrib/migrate-logs` here instead of deleting this attribute.

**Review by azahi — CHANGES_REQUESTED** [review]

**Comment by laalsaas** [comment]
May I ask why? When soju upstream adds some new command in a future release, we will miss that too. At keast if update doesn't consider that (which will probably happen, even more when the update is done by the bot)

**Comment by azahi** [comment]
> May I ask why? 

Because we don't need to build targets other than executables that will actually be actually used by the end user. Some packages may include cmds for testing, the notable example is [werf](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/networking/cluster/werf/default.nix#L26) which has [one](https://github.com/werf/werf/tree/main/cmd/buildah-test) one for checking buildah compatibility. We only need to look for and build all possible targets in the check phase so that we can run all possible tests if they exist.

> When soju upstream adds some new command in a future release, we will miss that too.

This assumption goes both ways: what if an executable that will only be useful for development purposes will pop up in the new version? Updating or approving the update is done in good faith by the maintainer, this includes checking the changelog and looking up diffs to prevent this from happening.

**Comment by laalsaas** [comment]
@azahi I disagree.
1. On a project with five executables, five of them being useful for users, zero being just for development purposes, how big are the chances that a new executable just for development will be added?
2. Even more important: having an executable more than needed doesn't really harm/disturb anyone. However, *missing* an executable, in case of `sojudb` one that is necessary to setup the software, is disruptive to the user: It is a crappy UX.

> Updating or approving the update is done in good faith by the maintainer, this includes checking the changelog and looking up diffs to prevent this from happening.

I agree that this is how the process is supposed to go, however, this has already failed for *two* new executables. Relying on it for the third one doesn't seem a wise choice.

I hope I could convince you. However, since you're probably the one who decides whether this gets merged, just write „we agree to disagree“, and I'll amend, sparing us the further discussion.

**Comment by azahi** [comment]
> I hope I could convince you. However, since you're probably the one who decides whether this gets merged, just write „we agree to disagree“, and I'll amend, sparing us the further discussion.

No, it's not a blocker, just nitpicking on my part :). Let's go with your vision here. Please remove the `unset subPackages` part in the checkPhase then as well.

**Comment by azahi** [comment]
Result of `nixpkgs-review pr 256974` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>soju</li>
  </ul>
</details>

**Review by azahi — APPROVED** [review]

**Pull request merged**